### PR TITLE
Move the charmstore code to jujulib.

### PR DESCRIPTION
**Pull request by jaycee**

### CODE REVIEW

**Comment by jaycee** [comment]
QA:

Ensure searching for charms, viewing their details and deploying all work. Same for bundles.

**Comment by jaycee** [comment]
There is a fair bit of follow up work to do here. They are noted with TODOs and XXXs. Briefly:

* Single CB model
* Fixing the weird clone behavior we have apparently been relying on

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js [hunk 1]
```diff
@@ -152,11 +152,423 @@ var module = module;
...
+          if (i > 0) {
+            qs += '&';
+          }
+          qs += key;
+          if (value && value !== '') {
+            qs += '=' + value;
+          }
+        });
+      }
+      qs = qs +
```
Comment (jaycee) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: The above qs code replaces a single use of Y.Querystring.Stringify.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js [hunk 2]
```diff
@@ -152,11 +152,423 @@ var module = module;
...
+          copy = {};
+          obj[key].forEach(function(val ,i) {
+            copy[i] = val;
+          });
+        } else {
+          copy =
+              (typeof obj[key] === 'object' && obj[key] !== null) ?
+                  JSON.parse(JSON.stringify(obj[key])) : obj[key];
+        }
+        host[newKey] = copy;
```
Comment (jaycee) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: The above copy code replaces a use of Y.merge, as described in the XXX comment above.

File: jujugui/static/gui/src/app/app.js [hunk 3]
```diff
@@ -1069,10 +1069,22 @@ YUI.add('juju-gui', function(Y) {
...
+        var processEntity = function(entity) {
+          if (entity.entityType === 'charm') {
+            return new Y.juju.models.Charm(entity);
+          } else {
+            return new Y.juju.models.Bundle(entity);
+          }
+        };
+        this.set(
+            'charmstore',
+            new Charmstore(charmstoreURL, apiPath, bakery, processEntity));
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: The charmstore constructor probably shouldn't be a global

File: development.ini [hunk 4]
```diff
@@ -15,7 +15,7 @@ jujugui.combine = true
 
 # JEM/IdM connections
 jujugui.interactive_login = false
-jujugui.jem_url =
+jujugui.jem_url =
```
Comment (hatched) [inline development.ini]: whitespace ;)

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js [hunk 5]
```diff
@@ -152,11 +152,423 @@ var module = module;
...
+   * @param processEnity {function} A function to massage entity data into the
+   *    desired form (e.g. turning it into juju gui model objects.
+   * @returns {Object} A client object for making charmstore API calls.
+   */
+  function charmstore(url, apiVersion, bakery, processEntity) {
+    this.url = url;
+    this.version = apiVersion;
+    this.bakery = bakery;
+
+    // XXX jcsackett 2015-11-09 Methods that return entity data should
```
Comment (hatched) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: card please

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js [hunk 6]
```diff
@@ -152,11 +152,423 @@ var module = module;
...
+      }
+      Object.keys(obj).forEach(function(key) {
+        // An exclude of 0 means "don't lowercase this level".
+        var newKey = key;
+        if (exclude !== 0) {
+          newKey = key.toLowerCase();
+        }
+        // Create shallow copies of objects for the copy.
+        var copy;
+        // XXX jcsackett 2015-11-09: Previously we used Y.merge to make shallow
```
Comment (hatched) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: card please

File: jujugui/static/gui/src/test/index.html [hunk 7]
```diff
@@ -12,7 +12,7 @@
 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
 SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
-General Public License for more details.
+General Public License for more details.
```
Comment (hatched) [inline jujugui/static/gui/src/test/index.html]: whitespace

File: jujugui/static/gui/src/test/index.html [hunk 8]
```diff
@@ -78,6 +78,7 @@
   <script src="/combo?app/assets/javascripts/bind-function-pollyfill.js&app/assets/javascripts/react-with-addons.js&app/assets/javascripts/react-dom.js&app/assets/javascripts/classnames.js&app/assets/javascripts/react-onclickoutside.js"></script>
   <script src="/combo?app/assets/javascripts/handlebars.runtime.js&app/components/templates.js&app/components/helpers.js"></script>
   <script src="/combo?app/assets/javascripts/yui/yui/yui.js&app/assets/javascripts/yui/loader/loader.js&app/assets/javascripts/d3.js"></script>
+  <script src="/combo?app/assets/javascripts/jujulib/juju.js"></script>
```
Comment (hatched) [inline jujugui/static/gui/src/test/index.html]: could this be combo'd with another to reduce requests? Maybe the YUI one?

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/test-jujulib-charmstore.js [hunk 9]
```diff
@@ -1,71 +1,33 @@
...
+      sendGetRequest: sinon.stub()
+    }
+    charmstore = new window.jujulib.charmstore('local/', 'v4', bakery);
   });
 
   afterEach(function() {
     charmstore = null;
   });
 
+
```
Comment (hatched) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/test-jujulib-charmstore.js]: extra line :)

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3507/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3508/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3509/
Test FAILed.

File: development.ini (hunk 4 above)
Comment (jaycee) [inline development.ini]: Fixed.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js (hunk 5 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: Added.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js (hunk 6 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: Added.

File: jujugui/static/gui/src/test/index.html (hunk 7 above)
Comment (jaycee) [inline jujugui/static/gui/src/test/index.html]: Fixed.

File: jujugui/static/gui/src/test/index.html (hunk 8 above)
Comment (jaycee) [inline jujugui/static/gui/src/test/index.html]: Done.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/test-jujulib-charmstore.js (hunk 9 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/test-jujulib-charmstore.js]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3511/
Test PASSed.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js [hunk 10]
```diff
@@ -152,11 +152,423 @@ var module = module;
...
+   *
+   * Provides access to the charmstore API.
+   */
+
+  /**
+   * Initializer
+   *
+   * @function charmstore
+   * @param url {String} The URL, including scheme and port, of the charmstore
+   * @param version {String} The api version, e.g. v4
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: apiVersion

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js [hunk 11]
```diff
@@ -152,11 +152,423 @@ var module = module;
...
+        models.push(entityData);
+      }, this);
+      successCallback(models);
+    },
+
+    /**
+      Recursively converts all keys to lowercase when assigning them to the
+      supplied host object.
+
+      @method lowerCaseKeys
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: _lowerCaseKeys

**Comment by huwshimi** [comment]
Great stuff @jaycee! QA OK :+1:

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js (hunk 10 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: Fixed.

File: jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js (hunk 11 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js]: Fixed.

**Comment by jaycee** [comment]
Thanks for the reviews, @huwshimi  and @hatched.

**Comment by jaycee** [comment]
Got a +1 from @hatched in IRC yesterday. :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3527/
Test PASSed.

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/1199

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**